### PR TITLE
oneAPI interfaces for DBSCAN

### PR DESCRIPTION
**Pull request by amgrigoriev**
-

### CODE REVIEW

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    const auto daal_data = interop::convert_to_daal_table<Float>(data);
+    const auto daal_weights = interop::convert_to_daal_table<Float>(weights);
+
+    array<int> arr_responses = array<int>::empty(row_count * 1);
+    array<int> arr_cluster_count = array<int>::empty(1);
+
+    const auto daal_responses = interop::convert_to_daal_homogen_table(arr_responses, row_count, 1);
+    const auto daal_core_observation_indices = interop::empty_daal_homogen_table<int>(1);
+    const auto daal_core_observations = interop::empty_daal_homogen_table<Float>(column_count);
+    const auto daal_cluster_count = interop::convert_to_daal_homogen_table(arr_cluster_count, 1, 1);
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp:79]: Why do we have different strategies for memory allocation, for `daal_responses`  memory is allocated via array, but for `daal_core_observations` via old numeric tables.

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    daal_dbscan::Parameter par(epsilon, dal::detail::integral_cast<std::size_t>(min_observations));
+    par.memorySavingMode = desc.get_mem_save_mode();
+    par.resultsToCompute = daal_dbscan::computeCoreIndices | daal_dbscan::computeCoreObservations;
+
+    const auto daal_data = interop::convert_to_daal_table<Float>(data);
+    const auto daal_weights = interop::convert_to_daal_table<Float>(weights);
+
+    array<int> arr_responses = array<int>::empty(row_count * 1);
+    array<int> arr_cluster_count = array<int>::empty(1);
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp:60]: Shouldn't we allocate device memory?

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp (hunk 1 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp:79]: daal_core_observations might be resized inside the kernel - pointer to the array might become invalid

File: cpp/oneapi/dal/detail/error_messages.hpp [hunk 3]
```diff
@@ -239,6 +239,10 @@ class ONEDAL_EXPORT error_messages {
     MSG(tau_leq_zero);
     MSG(epsilon_lt_zero);
     MSG(unknown_kernel_function_type);
+
+    /* DBSCAN */
+    MSG(weight_dimension_doesnt_match_data_dimension);
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/error_messages.hpp:251]: Replaced with existing message

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+using daal_dbscan_default_dense_t =
+    daal_dbscan::internal::DBSCANBatchKernel<Float, daal_dbscan::defaultDense, Cpu>;
+
+template <typename Float, daal::CpuType Cpu>
+class dbscan_compute_wrapper {
+public:
+    template <typename... Args>
+    auto compute(Args... args) {
+        auto pack = std::tuple<Args...>(args...);
+        const daal_dbscan::Parameter* par = std::get<sizeof...(Args) - 1>(pack);
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    auto core_observation_indices =
+        interop::convert_from_daal_homogen_table<int>(daal_core_observation_indices);
+    auto core_observations =
+        interop::convert_from_daal_homogen_table<Float>(daal_core_observations);
+    array<int> arr_core_flags = array<int>::full(row_count * 1, 0);
+    if (core_observation_indices.get_row_count() > 0) {
+        auto index_block = row_accessor<const int>(core_observation_indices).pull({ 0, -1 });
+        for (int index = 0; index < core_observation_indices.get_row_count(); index++) {
+            arr_core_flags.get_mutable_data()[index_block[index]] = 1;
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        }
+    }
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    return compute_result<Task>()
+        .set_responses(
+            dal::detail::homogen_table_builder{}.reset(arr_responses, row_count, 1).build())
+        .set_core_flags(
+            dal::detail::homogen_table_builder{}.reset(arr_core_flags, row_count, 1).build())
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    if (core_observation_indices.get_row_count() > 0) {
+        auto index_block = row_accessor<const int>(core_observation_indices).pull({ 0, -1 });
+        for (int index = 0; index < core_observation_indices.get_row_count(); index++) {
+            arr_core_flags.get_mutable_data()[index_block[index]] = 1;
+        }
+    }
+
+    return compute_result<Task>()
+        .set_responses(
+            dal::detail::homogen_table_builder{}.reset(arr_responses, row_count, 1).build())
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp (hunk 2 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp]: Added allocation of shared USM: device one caused problem in DAAL kernel (inside context.fill)

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+template <typename Float, daal::CpuType Cpu>
+using daal_dbscan_default_dense_t =
+    daal_dbscan::internal::DBSCANBatchKernel<Float, daal_dbscan::defaultDense, Cpu>;
+
+template <typename Float, daal::CpuType Cpu>
+class dbscan_compute_wrapper {
+public:
+    template <typename... Args>
+    auto compute(Args... args) {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp (hunk 2 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp (hunk 1 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp:76]: Due to that, it useless to allocate memory before the kernel call.

Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp:76]: Added a comment to clarify the difference

File: cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/table/row_accessor.hpp"
+
+namespace oneapi::dal::dbscan::backend {
+
+inline array<int> fill_core_flags(const table& core_observation_indices, std::int64_t row_count) {
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp:23]: `int` -> `int32_t`

File: cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    array<int> arr_core_flags = array<int>::full(row_count * 1, 0);
+    if (core_observation_indices.get_row_count() > 0) {
+        auto index_block = row_accessor<const int>(core_observation_indices).pull();
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp:26]: You can introduce function for the host similar to this:
https://github.com/oneapi-src/oneDAL/blob/9a8322c09c1414ac7c7c76dc825f32c0c5c5499e/cpp/oneapi/dal/backend/primitives/utils.hpp#L27

File: cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp (hunk 10 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp (hunk 11 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp]: Discussed offline

File: cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        .set_core_flags(dal::homogen_table::wrap(arr_core_flags, row_count, 1))
+        .set_core_observation_indices(core_observation_indices)
+        .set_core_observations(core_observations)
+        .set_cluster_count(static_cast<std::int64_t>(arr_cluster_count[0]));
+}
+
+template <typename Float, typename Task>
+static compute_result<Task> compute(const context_cpu& ctx,
+                                    const descriptor_t& desc,
+                                    const compute_input<Task>& input) {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/cpu/compute_kernel_brute_force.cpp]: Done

File: cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace oneapi::dal::dbscan::backend {
+
+inline array<std::int32_t> fill_core_flags(const table& core_observation_indices,
+                                           std::int64_t row_count) {
+    array<std::int32_t> arr_core_flags = array<std::int32_t>::full(row_count * 1, 0);
+    if (core_observation_indices.get_row_count() > 0) {
+        auto index_block = row_accessor<const std::int32_t>(core_observation_indices).pull();
+        auto arr_core_flags_ptr = arr_core_flags.get_mutable_data();
+        for (std::int32_t index = 0; index < core_observation_indices.get_row_count(); index++) {
```
Comment (makart19) [inline cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp:29]: types mismatch std::int32_t index vs core_observation_indices.get_row_count() (aka std::int64_t), this loop can cause index overflow

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    const int64_t min_observations = desc.get_min_observations();
+
+    daal_dbscan::Parameter par(epsilon, dal::detail::integral_cast<std::size_t>(min_observations));
+    par.memorySavingMode = desc.get_mem_save_mode();
+    par.resultsToCompute = daal_dbscan::computeCoreIndices | daal_dbscan::computeCoreObservations;
+
+    const auto daal_data = interop::convert_to_daal_table(queue, data);
+    const auto daal_weights = interop::convert_to_daal_table(queue, weights);
+
+    array<int> arr_responses = array<int>::empty(queue, row_count * 1, sycl::usm::alloc::device);
```
Comment (makart19) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp:61]: int -> std::int32_t

File: cpp/oneapi/dal/algo/dbscan/common.hpp [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+template <typename Task>
+constexpr bool is_valid_task_v = dal::detail::is_one_of_v<Task, task::clustering>;
+
+template <typename Task = task::by_default>
+class descriptor_base : public base {
+    static_assert(is_valid_task_v<Task>);
+
+public:
+    using tag_t = descriptor_tag;
+    using float_t = float;
```
Comment (makart19) [inline cpp/oneapi/dal/algo/dbscan/common.hpp:73]: remove

File: cpp/oneapi/dal/algo/dbscan/compute_types.cpp [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+template <typename Task>
+void compute_result<Task>::set_core_observations_impl(const table& value) {
+    impl_->core_observations = value;
+}
+
+template <typename Task>
+void compute_result<Task>::set_cluster_count_impl(std::int64_t value) {
+    if (value < 0) {
+        //        throw domain_error(dal::detail::error_messages::cluster_count_lt_zero());
```
Comment (makart19) [inline cpp/oneapi/dal/algo/dbscan/compute_types.cpp:123]: add exception ?

File: cpp/oneapi/dal/algo/dbscan/test/batch.cpp [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+    SKIP_IF(this->not_float64_friendly());
+    using Float = std::tuple_element_t<0, TestType>;
+
+    const te::dataframe data = GENERATE_DATAFRAME(
+        te::dataframe_builder{ "workloads/road_network/dataset/road_network_20t_cluster.csv" });
+    const table x = data.get_table(this->get_policy(), this->get_homogen_table_id());
+
+    const double epsilon = 1.0e3;
+    const std::int64_t min_observations = 220;
+    constexpr Float ref_dbi = 0.00036f;
```
Comment (makart19) [inline cpp/oneapi/dal/algo/dbscan/test/batch.cpp:340]:
```suggestion
    constexpr Float ref_dbi = Float(0.00036);
```

File: cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp (hunk 13 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/fill_core_flags.hpp]: it should be int64_t - it is just filling of the array. Fixed.

File: cpp/oneapi/dal/algo/dbscan/common.hpp (hunk 15 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/common.hpp]: Done

File: cpp/oneapi/dal/algo/dbscan/compute_types.cpp (hunk 16 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/compute_types.cpp]: Removed the "if" block. It's a separate discussion whether to check the value or not.

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp (hunk 14 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp]: Done

**Review by makart19 — APPROVED** [review]

**Pull request merged**